### PR TITLE
[CHORE]: move utf8 functions from daft-dsl to daft-functions 

### PR DESCRIPTION
**Pull request by ConeyLiu**
This refers to #2854 moving the UTF-8 functions from daft-dsl to daft-functions

### CODE REVIEW

File: src/daft-sql/src/modules/utf8.rs [hunk 1]
```diff
@@ -1,313 +1,509 @@
...
+            }
+
+            fn arg_names(&self) -> &'static [&'static str] {
+                &[$arg_name_1, $arg_name_2]
+            }
+        }
+    };
+}
+
+macro_rules! utf8_function_three_arguments {
```
Comment (ConeyLiu) [inline src/daft-sql/src/modules/utf8.rs:94]: Three macros are ugly, I haven't been able to find better solutions. I would appreciate any assistance or suggestions.

Comment (universalmind303) [reply src/daft-sql/src/modules/utf8.rs:94]: I think if we created a few helper functions: 

```rs
fn utf8_unary(
    func: impl Fn(ExprRef) -> ExprRef,
    sql_name: &str,
    arg_name: &str,
    inputs: &[sqlparser::ast::FunctionArg],
    planner: &crate::planner::SQLPlanner,
) -> SQLPlannerResult<ExprRef> {
    match inputs {
        [input] => {
            let input = planner.plan_function_arg(input)?;
            Ok(func(input))
        }
        _ => invalid_operation_err!(
            "invalid arguments for {sql_name}. Expected {sql_name}({arg_name})",
        ),
    }
}

fn utf8_binary(...) 
fn utf3_ternary(...)
```

that'd allow us to simplify the macros into a single one pretty easily. It also pulls the bulk of the code out of macro which will cut down on code generation during compiling. 

with the helper functions, you can create a single macro that matches on all 3 patterns 
```rs
macro_rules! utf8_function {
    ($name:ident, $sql_name:expr, $func:expr, $doc:expr, $arg_name:expr) => {
        // ...
        impl SQLFunction for $name {
            fn to_expr(..) -> SQLPlannerResult<ExprRef> {
                utf8_unary($func, $sql_name, $arg_name, inputs, planner)
            }
            // ...
        }
    };
    ($name:ident, $sql_name:expr, $func:expr, $doc:expr, $arg_name_1:expr, $arg_name_2:expr) => {
        // ...
        impl SQLFunction for $name {
            fn to_expr(..) -> SQLPlannerResult<ExprRef> {
                utf8_binary($func, $sql_name, $arg_name_1, $arg_name_2, inputs, planner)
            }
            // ...
        }
    };
    ($name:ident, $sql_name:expr, $func:expr, $doc:expr, $arg_name_1:expr, $arg_name_2:expr, $arg_name_3:expr) => {
        // ...
        impl SQLFunction for $name {
            fn to_expr(..) -> SQLPlannerResult<ExprRef> {
                utf8_ternary($func, $sql_name, $arg_name_1, $arg_name_2, $arg_name_3, inputs, planner)
            }
            // ...
        }
    };
}
```

Comment (ConeyLiu) [reply src/daft-sql/src/modules/utf8.rs]: Cool, I have updated it based on your suggestions.

File: src/daft-sql/src/modules/utf8.rs [hunk 2]
```diff
@@ -1,313 +1,517 @@
...
+        parent.add_fn("ends_with", SQLUtf8EndsWith);
+        parent.add_fn("starts_with", SQLUtf8StartsWith);
+        parent.add_fn("contains", SQLUtf8Contains);
+        parent.add_fn("split", SQLUtf8Split);
         // TODO add split variants
         // parent.add("split", f(Split(false)));
-        parent.add_fn("regexp_match", Match);
-        parent.add_fn("regexp_extract", Extract(0));
-        parent.add_fn("regexp_extract_all", ExtractAll(0));
-        parent.add_fn("regexp_replace", Replace(true));
```
Comment (ConeyLiu) [inline src/daft-sql/src/modules/utf8.rs:40]: @universalmind303 I think this should be a bug. The `regexp_replace` set the `regexp` to True, while the [implementation](https://github.com/Eventual-Inc/Daft/pull/3101/files#diff-06e4a592259b845ed2b29b25e79451299bf1bd8947755f6752bd895610bfe88fL209) is not.

**Comment by ConeyLiu** [comment]
@universalmind303 could you help to review this again? Thanks a lot.

**Comment by universalmind303** [comment]
> @universalmind303 could you help to review this again? Thanks a lot.

sorry for the delay. Ill take a look at it today!

**Review by universalmind303 — APPROVED** [review]
@ConeyLiu Looks good! Thanks for this! Once the conflicts are resolved, I can go ahead and merge this in!

**Comment by ConeyLiu** [comment]
Thanks @universalmind303 for your time. I have rebased it.

File: daft/dataframe/dataframe.py [hunk 3]
```diff
@@ -1601,7 +1601,7 @@ def limit(self, num: int) -> "DataFrame":
             │ ---   │
             │ Int64 │
             ╞═══════╡
-            │ 1     │ f
```
Comment (ConeyLiu) [inline daft/dataframe/dataframe.py:1604]: A small fix since this breaks the doc tests.

**Pull request merged**